### PR TITLE
Divulgence crash tests [DPP-433]

### PR DESCRIPTION
**Pull request by kamil-da**
#### Problematic scenario
1. Party `Alice` is hosted on Participant `P1`. Party `Bob` is hosted on participant `P2`.
2. Both `Alice` and `Bob` are signatories on a shared contract `C1`.
3. `Alice` exercises a non-consuming choice on `C1` that creates a contract `C2` with contract key `CK` with only `Alice` as a stakeholder in transaction `T1`.
4. Due to `Bob` being a stakeholder on `C1`, he sees the full transaction `T1`, including the creation of `C2`.
5. The participant `P2` stores `C2` with the hash of contract key `CK` in the participant_contracts table.
6. Alice archives contract `C2`. But since `Bob` is not a stakeholder on `C2`, he doesn't learn about the archival.
7. Any subsequently created contract `C3` with the same contract key `CK` and witnessed by `Bob` (irrespective of being a stakeholder or via divulgence), will cause an error in participant `P2`.

#### Changes
- A test case exposing the problematic scenario added to the `ledger-api-test-tool`

Confirmed that the test fails on ledgers without the append-only schema.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (kamil-da) [inline ledger/test-common/src/main/daml/model/Test.daml:718]: This will be used in a subsequent PR.

**Comment by rautenrieth-da** [comment]
Why do all tests pass? Don't we expect the new test to fail on all non-appendonly index dbs?

**Comment by gerolf-da** [comment]
They pass, because they would only fail in a privacy-aware multi-participant setup, which we cannot run in this repository.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:378]: Maybe check that the contract is usable for interpretation here?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:374]: I think the other direction is interesting as well where this command is submitted on ledger 2, see https://gist.github.com/cocreature/c134d2ba0dd974708ed061a5ec053256 for an example. They both run into the same error at the moment but I think it’s helpful to test for both.

Comment (cocreature) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:374]: If you want to be extra thorough test the cases where the first gets created on p1 and the second on p2 and the other way around.

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (cocreature) [reply ledger/test-common/src/main/daml/model/Test.daml:718]: Is that to test that the same thing works for divulgence instead of disclosure? Because if not, we should definitely have a testcase for this even if it happens to not be broken at the moment.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:374]: You're right, it's worth testing the second case. Actually I'm currently working on it - @mziolekda sent me the scenario.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:378]: Do you mean checking the structure of the Scala object we get? Or fetching the contract to be sure that it's ok?

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (kamil-da) [reply ledger/test-common/src/main/daml/model/Test.daml:718]:
```
  p1 <- allocatePartyOn "p1" (ParticipantName "A")
  p2 <- allocatePartyOn "p2" (ParticipantName "B")
  helper <- submit p1 $ createCmd (Helper p1 p2)
  cid <- submit p1 $ exerciseCmd helper (DiscloseCreate p1)
  -- fetch f1 by p2 should succeed
  submit p1 $ archiveCmd cid
  -- fetch f2 by p2 should fail
  ```

  That's the scenario that's going to be implemented.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (cocreature) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:378]: Submit a command on p2 that fetches the contract to check that it has been disclosed.

Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:378]: Ok, thanks!

Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:379]: @cocreature I've added additional verification that the disclosed contract can be fetched by the `party2`. I would appreciate your review :)

Comment (cocreature) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:379]: Good stuff!

**Review by cocreature — APPROVED** [review]
lgtm, I’ll leave it up to you/your team whether you want to handle the divulgence and the flipped participant cases in this PR or separately.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:379]: Thanks!

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:386]: It would be interesting to fetch again here, to verify that the submission is rejected on-ledger in a failed completion. So even if `p2`'s participant thinks the contract is still active, the committer must reject the transaction.

Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:386]: That was going to be a part of another test case but I can make this check here as well.

Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:386]: Done.

**Review by mziolekda — APPROVED** [review]
Great work @kamil-da! Thanks!
Hold off with merging until we patch the indexer though.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:388]:
```suggestion
        _ <- ledger1.exercise(party1, withKey1.exerciseArchive(_))
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:431]:
```suggestion
        _ <- ledger1.exercise(party1, withKey1.exerciseArchive(_))
```

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (cocreature) [inline ledger/test-common/src/main/daml/model/Test.daml:696]:
```suggestion
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala]: Fixed in #9970 and merged into this PR.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala]: Fixed in #9970 and merged into this PR.

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (kamil-da) [reply ledger/test-common/src/main/daml/model/Test.daml]: Fixed in #9970 and merged into this PR.

**Pull request merged**
